packages/contracts/contracts/EpochKeyProofVerifier.sol

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IUnirep} from './interfaces/IUnirep.sol';
+import {IVerifier} from './interfaces/IVerifier.sol';
+
+contract EpochKeyProofVerifier is IUnirep {
+    IVerifier public immutable epochKeyVerifier;
+
+    uint256 public constant SNARK_SCALAR_FIELD =
+        21888242871839275222246405745257275088548364400416034343698204186575808495617;
+
+    constructor(IVerifier _epochKeyVerifier) {
+        epochKeyVerifier = _epochKeyVerifier;
+    }
+
+    function decodeEpochKeyControl(
+        uint256 control
+    )
+        public
+        pure
+        returns (
+            uint256 revealNonce,
+            uint256 attesterId,
+            uint256 epoch,
+            uint256 nonce
+        )
+    {
+        revealNonce = (control >> 232) & 1;
+        attesterId = (control >> 72) & ((1 << 160) - 1);
+        epoch = (control >> 8) & ((1 << 64) - 1);
+        nonce = control & ((1 << 8) - 1);
+        return (revealNonce, attesterId, epoch, nonce);
+    }
+
+    function decodeEpochKeySignals(
+        uint256[] calldata publicSignals
+    ) public pure returns (EpochKeySignals memory) {
+        EpochKeySignals memory signals;
+        signals.epochKey = publicSignals[0];
+        signals.stateTreeRoot = publicSignals[1];
+        signals.data = publicSignals[3];
+        // now decode the control values
+        (
+            signals.revealNonce,
+            signals.attesterId,
+            signals.epoch,
+            signals.nonce
+        ) = decodeEpochKeyControl(publicSignals[2]);
+        return signals;
+    }
+
+    function verifyEpochKeyProof(
```

Comment:
This function could probably be named more generically. In the calling contract it will likely be stored as a variable like:

```solidity
EpochKeyProofVerifier epochKeyVerifier;
```

If the function is named `verifyEpochKeyProof` you get

```solidity
epochKeyVerifier.verifyEpochKeyProof(...);
```

Instead it's okay to name it like `verifyAndCheck`. Then all the verifiers can have similar function names. In the calling code it should look like

```solidity
epochKeyVerifier.verifyAndCheck(...);
```

Reply:
Refactored